### DISCUSSION TITLE
When to use `eventId` and/or idempotency key when creating a message

### DISCUSSION THREAD

**Discussion by mderriey** [svix/svix-webhooks#1251] [Q&A] [answered]
Hi there

We're having trouble understanding when to use [an idempotency key](https://docs.svix.com/idempotency) and/or [an event id](https://api.svix.com/docs#tag/Message/operation/v1.message.create) when creating a message.

We understood this from the docs:

- If the same event id is encountered within 24 hours, a 409 is returned.
- If the same idempotency key is encountered within 12 hours, then the original status code and body are returned.

While the behaviors are different, it sounds to us like they serve the same purpose, which is to avoid duplicates being processed on the Svix side.

Do you have any insights as to when to use one over the other, or how they compliment each other in some way?

Cheers

**Comment by mderriey** [comment]
Friendly bump in case this was missed. No pressure, but I posted a question at roughly the same time as this one and it was answered super quickly, which is greatly appreciated by the way 😉

**Reply by tasn** [reply, marked as answer]
It was missed, thanks for bumping!

More info:
1. `eventId`: this is a unique identifier per message. It's meant to be used if you already have a unique internal identifiers for events. The uniqueness in enforced over a 24hr period for performance reasons, but this should be unique per application. You can query by this identify later as well to correlate webhooks with your internal event IDs. Sending two events with the same ID is therefore an error, as those should be unique.
2. Idempotency: this is intended to be a mechanism to allow to safely retry API calls (this applies to other calls, not just create message). Consider this scenario: you make an API call and get a timeout (but it actually succeeded in the background) so you retry. You now want the second call to return a success, not to return an error, because with retries all you care about really is that the call is successful and you want to get the returned data. This is why idempotency just returns the same data and not an error.

So in short: if you want retry safety use idempotency, if you happen to have internal event IDs correlated with webhooks, set the `eventId`.

I hope this clarifies things.

**Reply by mderriey** [reply]
I see, thanks a lot.

Do you know which behavior kicks in first if a message with an existing `eventId` and an already-seen idempotency key is received?

**Reply by tasn** [reply]
Idempotency.

**Reply by mderriey** [reply]
Sweet, thanks a lot!
